src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs

```diff
@@ -45,6 +45,19 @@ public SocketInput(MemoryPool memory, IThreadPool threadPool, IBufferSizeControl
 
         public bool IsCompleted => ReferenceEquals(_awaitableState, _awaitableIsCompleted);
 
+        public bool CheckFinOrThrow()
```

Comment:
Should `RemoteIntakeFin` be public now? Can we make it tri-state to avoid the additional locking?